packages/core/contracts/trader/uniswap-broker/UniswapBroker.sol

```diff
@@ -0,0 +1,171 @@
+     * @param to recipient of the trade proceeds.
+     * @param to deadline to limit when the trade can execute. If the tx is mined after this timestamp then revert.
+     */
+
+    function swapToPrice(
+        address uniswapRouter,
+        address uniswapFactory,
+        address tokenA,
+        address tokenB,
+        uint256 truePriceTokenA,
+        uint256 truePriceTokenB,
+        uint256 maxSpendTokenA,
+        uint256 maxSpendTokenB,
+        address to,
+        uint256 deadline
+    ) public {
+        IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+
+        // true price is expressed as a ratio, so both values must be non-zero
+        require(truePriceTokenA != 0 && truePriceTokenB != 0, "SwapToPrice: ZERO_PRICE");
+        // caller can specify 0 for either if they wish to swap in only one direction, but not both
+        require(maxSpendTokenA != 0 || maxSpendTokenB != 0, "SwapToPrice: ZERO_SPEND");
+
+        bool aToB;
+        uint256 amountIn;
+        {
+            (uint256 reserveA, uint256 reserveB) = getReserves(uniswapFactory, tokenA, tokenB);
+            (aToB, amountIn) = computeTradeToMoveMarket(truePriceTokenA, truePriceTokenB, reserveA, reserveB);
+        }
+
+        require(amountIn > 0, "SwapToPrice: ZERO_AMOUNT_IN");
+
+        // spend up to the allowance of the token in
+        uint256 maxSpend = aToB ? maxSpendTokenA : maxSpendTokenB;
+        if (amountIn > maxSpend) {
+            amountIn = maxSpend;
+        }
+
+        address tokenIn = aToB ? tokenA : tokenB;
+        address tokenOut = aToB ? tokenB : tokenA;
+        TransferHelper.safeTransferFrom(tokenIn, msg.sender, address(this), amountIn);
+        TransferHelper.safeApprove(tokenIn, address(router), amountIn);
+
+        address[] memory path = new address[](2);
+        path[0] = tokenIn;
+        path[1] = tokenOut;
+
+        router.swapExactTokensForTokens(
+            amountIn,
+            0, // amountOutMin: we can skip computing this number because the math is tested
+            path,
+            to,
+            deadline
+        );
+    }
+
+/**
+     * @notice Given the "true" price a token (represented by truePriceTokenA/truePriceTokenB) and the reservers in the
+     * uniswap pair, calculate: a) the direction of trade (aToB) and b) the amount needed to trade (amountIn) to move
+     * the pool price to be equalto the true price.
+     * @dev Note that this method uses the Babylonian square root method which has a small margin of error which will
+     * result in a small over or under estimation on the size of the trade needed.
+     * @param truePriceTokenA the nominator of the true price.
+     * @param truePriceTokenB the denominatornominator of the true price.
+     * @param reserveA number of token A in the pair reserves
+     * @param reserveB number of token B in the pair reserves
+     */
+//
+    function computeTradeToMoveMarket(
```

Comment:
Can you explain the differences in goal/implementation between this method and the uniswap method here: https://github.com/Uniswap/uniswap-v2-periphery/blob/dda62473e2da448bc9cb8f4514dadda4aeede5f4/contracts/libraries/UniswapV2LiquidityMathLibrary.sol#L17?

Reply:
sure. In their implementation the method tries to compute the profit maximising trade. In my method, it tries to get the price as close as posible to the goal price. This difference is in theirs the fees of the pool are considered while in mine it is not. In playing with their contract, it would always result in an error of a few percent below the desired "true" price where my equation will take the market to the exact market rate.

does that make sence?